src/ui/messages.rs

```diff
@@ -0,0 +1,213 @@
+use pango;
+use gtk;
+use gtk::prelude::*;
+
+use ui::ui::HlDefs;
+use ui::font::{Font, FontUnit};
+use ui::color::Color;
+use nvim_bridge::{MsgShow, MsgShowKind};
+
+struct Message {
+    container: gtk::Box,
+}
+
+impl Message {
+    fn new(msg: &MsgShow, hl_defs: &HlDefs, css_provider: &gtk::CssProvider, size: f64) -> Self {
+        let label = gtk::Label::new(None);
+
+        let mut content = String::new();
+
+        for chunk in msg.content.iter() {
+            let hl = hl_defs.get(&chunk.0).unwrap();
+            let markup = hl.pango_markup(
+                &chunk.1,
+                &hl_defs.default_fg,
+                &hl_defs.default_bg,
+                &hl_defs.default_sp,
+            );
+
+            content += &markup;
+        }
+
+        label.set_markup(&content);
+        label.set_halign(gtk::Align::Start);
+        label.set_line_wrap(true);
+        label.set_line_wrap_mode(pango::WrapMode::WordChar);
+        label.set_xalign(0.0);
+
+        let box_ = gtk::Box::new(gtk::Orientation::Horizontal, 0);
+        box_.set_halign(gtk::Align::End);
+        box_.set_valign(gtk::Align::Start);
+
+        let buf = get_icon_pixbuf(&msg.kind, &hl_defs.default_fg, size);
+        let kind = gtk::Image::new_from_pixbuf(&buf);
+
+        box_.pack_start(&kind, false, true, 0);
+        box_.pack_start(&label, false, true, 0);
+
+        add_css_provider!(css_provider, box_, label, kind);
+
+        Self {
+            container: box_,
+        }
+    }
+
+    fn widget(&self) -> gtk::Widget {
+        self.container.clone().upcast::<gtk::Widget>()
+    }
+}
+
+impl Drop for Message {
+    fn drop(&mut self) {
+        self.container.destroy();
+    }
+}
+
+pub struct MessagesHandler {
+    /// Our css provider.
+    css_provider: gtk::CssProvider,
+    /// Container where our message widegts will live.
+    container: gtk::Box,
+
+    messages: Vec<Message>,
+
+    font: Font,
+}
+
+impl MessagesHandler {
+    pub fn new(parent: &gtk::Overlay) -> Self {
+        let css_provider = gtk::CssProvider::new();
+
+        let container = gtk::Box::new(gtk::Orientation::Vertical, 0);
+        container.set_valign(gtk::Align::Start);
+
+        parent.add_overlay(&container);
+        parent.set_overlay_pass_through(&container, true);
+
+        // Work around some intial draw issues.
+        container.show_all();
+        container.hide();
+
+        MessagesHandler {
+            css_provider,
+            container,
+            messages: vec!(),
+            font: Font::default(),
+        }
+    }
+
+    pub fn show(&mut self, msg: &MsgShow, hl_defs: &HlDefs) {
+
+        if msg.replace_last {
+            self.messages.pop();
+        }
+
+        let msg = Message::new(msg, hl_defs, &self.css_provider, self.font.height as f64);
+        self.container.pack_end(&msg.widget(), false, true, 5);
```

Comment:
Is the last parameter of `self.container.pack_end` here meant to be `5`? When set to 5, there is a small bit of padding between the GNvim window and the message (notice the gap between the top of the message and the window):
![image](https://user-images.githubusercontent.com/46855713/60058610-a815fd00-96ae-11e9-98be-7f4f7e22ecd0.png)

With `0` used instead, it looks like this, with no gap between the message and window, which I think is a bit cleaner:
![image](https://user-images.githubusercontent.com/46855713/60058616-aba98400-96ae-11e9-8f29-33cbc949178d.png)

If we wanted to keep the padding, we might want to add some to the right too (although that might look strange, haven't tried it).

Reply:
I added the padding there so there would indeed be some spacing between the messages.